recipes/gmp/all/test_package/conanfile.py

```diff
@@ -1,22 +1,35 @@
-from conans import ConanFile, CMake, tools
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
 import os
 
 
 class TestPackageConan(ConanFile):
-    settings = "os", "compiler", "build_type", "arch"
-    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "CMakeDeps", "VirtualRunEnv"
+    test_type = "explicit"
+
+    def layout(self):
+        cmake_layout(self)
+
+    def requirements(self):
+        self.requires(self.tested_reference_str)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["ENABLE_CXX"] = self.dependencies["gmp"].options.enable_cxx
+        tc.variables["TEST_PIC"] = "fPIC" in self.dependencies["gmp"].options and self.dependencies["gmp"].options.fPIC
+        tc.generate()
 
     def build(self):
         cmake = CMake(self)
-        cmake.definitions["ENABLE_CXX"] = self.options["gmp"].enable_cxx
-        cmake.definitions["TEST_PIC"] = "fPIC" in self.options["gmp"] and self.options["gmp"].fPIC
         cmake.configure()
         cmake.build()
 
     def test(self):
-        if not tools.cross_building(self):
-            bin_path = os.path.join("bin", "test_package")
-            self.run(bin_path, run_environment=True)
+        if can_run(self):
+            bin_path = os.path.join(self.cpp.build.bindirs[0], "test_package")
+            self.run(bin_path, env="conanrun")
             if self.options["gmp"].enable_cxx:
```

Comment:
```suggestion
            if self.dependencies["gmp"].options.enable_cxx:
```

Reply:
self.dependencies is not defined in test(), it can't work. It's something which has already been experimented in conan-center.

Reply:
Hmm that's strange, I think it worked for me locally with conan 2.0.0-beta6 (and the existing code gave an error).

Reply:
It doesn't work in conan v1 and conan-center builds everything with conan 1.54.0 currently.

Reply:
Understood, thanks.